src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +26,96 @@ public static class Renamer
                 newName, optionSet, cancellationToken);
         }
 
+        /// <summary>
+        /// Similar to calling <see cref="Document.WithName(string)" /> with additional changes to the solution. 
+        /// Each change is added as a <see cref="RenameDocumentAction"/> in the returned <see cref="RenameDocumentActionSet.ApplicableActions" />.
+        /// 
+        /// Each action may individually encounter errors that prevent it from behaving correctly. Those are reported in <see cref="RenameDocumentAction.GetErrors(System.Globalization.CultureInfo?)"/>.
+        /// 
+        /// Current supported actions that may be returned: 
+        /// * <see cref="RenameSymbolDocumentAction"/> that will rename the type to match the document name
+        /// </summary>
+        public static async Task<RenameDocumentActionSet> RenameDocumentNameAsync(
+            Document document,
+            string newDocumentName,
+            OptionSet optionSet,
+            CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (newDocumentName == null)
+            {
+                throw new ArgumentNullException(nameof(newDocumentName));
+            }
+
+            var actions = new ArrayBuilder<RenameDocumentAction>();
+
+            if (!newDocumentName.Equals(document.Name))
+            {
+                var renameAction = await RenameSymbolDocumentAction.TryCreateAsync(document, newDocumentName, optionSet, cancellationToken).ConfigureAwait(false);
+
+                if (renameAction is object)
+                {
+                    actions.Add(renameAction);
+                }
+            }
+
+            return new RenameDocumentActionSet(
```

Comment:
why do we return an action set that may have no actions in it.  seems strange.

Reply:
If a consumer calls this API, we still want them to get the default `WithDocumentName` and `WithDocumentFolders` update (as appropriate for the API) even if no other actions should be taken. Since actions have a description, it's more indicative of not needing to ask the user of anything before applying actions.

Reply:
1. i don't see how WithDocmentFolders could apply here.  All the user provided was a newDocumentName.
2. i see the argument for why this is hte way it is.  We need to really flesh out the docs on RenameDoucmenetActionSet to make this clear.  i.e. that ApplicableActions may be empty, but the ActionSet still represents the ability to change the document name.

Reply:
I updated some documentation to help capture this nuance. 

For 1. I didn't want to break up into different logic for if the document name or folders need to be updated. It will be a no-op if they're the same.